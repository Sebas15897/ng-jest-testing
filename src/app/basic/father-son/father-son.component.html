<h5>Hijo</h5>
<code>
  {{ client | json }}
</code>

<button id="btn-change" class="btn btn-outline-primary" *ngIf="client" (click)="onChange(5)">
  Cambiar ID
</button>

<button id="btn-delete" class="btn btn-outline-danger" *ngIf="client" (click)="onDelete()">
  Borrar cliente
</button>
