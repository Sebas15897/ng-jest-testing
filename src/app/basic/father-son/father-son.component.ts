import { Component, EventEmitter, Input, Output } from '@angular/core';
import { IClient } from '../interfaces';

@Component({
  selector: 'app-father-son',
  templateUrl: './father-son.component.html',
  styleUrls: ['./father-son.component.css']
})
export class FatherSonComponent {
  @Input() client?: IClient;
  @Output() onDeleteClient = new EventEmitter();
  @Output() onClientUpdated = new EventEmitter<IClient>();


  onDelete() {
    this.client = undefined;
    this.onDeleteClient.emit();
  }

  onChange(id: number) {
    if (!this.client) return;
    this.client = { ...this.client, id };
    this.onClientUpdated.emit({ ...this.client });
  }
}
