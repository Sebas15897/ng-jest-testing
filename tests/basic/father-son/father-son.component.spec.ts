import { ComponentFixture, TestBed } from '@angular/core/testing';
import { FatherSonComponent } from '../../../src/app/basic/father-son/father-son.component';

describe('FatherSonComponent', () => {
  let component: FatherSonComponent;
  let fixture: ComponentFixture<FatherSonComponent>;
  let compilet: HTMLElement;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [FatherSonComponent],
    });
    fixture = TestBed.createComponent(FatherSonComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compilet = fixture.nativeElement;
    jest.clearAllMocks();
  });

  test('should create', () => {
    expect(component).toBeTruthy();
  });

  test('debe hacer match con el snapshot', () => {
    expect(compilet).toMatchSnapshot();
  });

  test('no deben de aparecer los botones si no hay un client', () => {
    const btn = compilet.querySelectorAll('button');
    expect(btn.length).toBeFalsy();
  });

  test('deben de aparecer los 2 botones si  hay un client', () => {
    component.client = { name: 'Juan', id: 1 };
    fixture.detectChanges();
    const buttons = compilet.querySelectorAll('button');
    expect(buttons.length).toBe(2);
  });

  test('Si hay cliente hacer match con el snapshot', () => {
    component.client = { name: 'Juan', id: 1 };
    fixture.detectChanges();
    expect(compilet).toMatchSnapshot();
  });

  test('Debe emitir  cuando se hace click en el boton de eliminar', () => {
    component.client = { name: 'Juan', id: 1 };
    fixture.detectChanges();
    jest.spyOn(component.onDeleteClient, 'emit');

    const btnDelete = compilet.querySelector('#btn-delete');
    btnDelete?.dispatchEvent(new Event('click'));
    expect(component.onDeleteClient.emit).toHaveBeenCalled();
  });

  test('Debe de emitir onClientUpdated con el botÃ³n de cambiar ID', () => {
    component.client = { name: 'Juan', id: 1 };
    fixture.detectChanges();
    jest.spyOn(component.onClientUpdated, 'emit');

    const btnChange = compilet.querySelector('#btn-change');
    btnChange?.dispatchEvent(new Event('click'));
    expect(component.onClientUpdated.emit).toHaveBeenCalledWith({
      name: 'Juan',
      id: 5,
    });
  });

  test('Debe de imitir onChangeCLient con el ID especificado si hay un cliente', () => {
    jest.spyOn(component.onClientUpdated, 'emit');
    component.onChange(10);
    expect(component.onClientUpdated.emit).not.toHaveBeenCalled();

    component.client = { name: 'Juan', id: 1 };
    fixture.detectChanges();
    component.onChange(10);
    expect(component.onClientUpdated.emit).toHaveBeenCalledWith({
      name: 'Juan',
      id: 10,
    });
  });
});
