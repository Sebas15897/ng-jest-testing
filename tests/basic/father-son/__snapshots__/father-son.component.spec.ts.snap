// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FatherSonComponent Si hay cliente hacer match con el snapshot 1`] = `
<div
  id="root4"
>
  <h5>
    Hijo
  </h5>
  <code>
     {
  "name": "Juan",
  "id": 1
}

  </code>
  <button
    class="btn btn-outline-primary"
    id="btn-change"
  >
     Cambiar ID

  </button>
  
  <button
    class="btn btn-outline-danger"
    id="btn-delete"
  >
     Borrar cliente

  </button>
  
</div>
`;

exports[`FatherSonComponent debe hacer match con el snapshot 1`] = `
<div
  id="root1"
>
  <h5>
    Hijo
  </h5>
  <code>
     

  </code>
  
  
</div>
`;
